Add table column to system_traces.sessions table

radoslawcybulski opened the pull request:
Add table name to tracing in alternator

Add a table name to Alternator's tracing output, as some clients would
like to consistently receive this information.

- add missing `tracing::add_table_name` in `executor::scan`
- add emiting tables' names in `trace_state::build_parameters_map`
- update tests, so when tracing is looked for it is filtered by table's
  name, which confirms table is being outputed.

Refs #26618
Fixes #24031

radoslawcybulski commented:
> Why no cover letter nor commit msg?

Long story short - it's placeholder, because i was migrating from laptop to desktop and finding branches on github is way more annoying that finding PRs.

nyh commented on test/alternator/test_tracing.py:
```diff
@@ -106,20 +106,22 @@ def iterate_tracing_sessions(dynamodb):
     # when looking for the other requests.
     global last_scan
     trace_sessions_table = dynamodb.Table('.scylla.alternator.system_traces.sessions')
+    expected_table = f'{{table : alternator_{table_name}.{table_name}}}'
```
Do we really want the table name listed to be the CQL name "alternator_xyz.xyz", and not the Alternator name, just "xyz"?
Why?
Again, we need to document this somewhere.

nyh replied:
Probably docs/dev/tracing.md (but please search if there's anything else)

The document might be missing other Alternator-specific information - I think we also have "alternator_op" that isn't documented.

In my opinion, if this is an Alternator command, users will expected the different parameters like "query", "alternator_op" and now "table" to be in Alternator table name, not the CQL name.

radoslawcybulski replied:
Updated table name and update documentation.

nyh reviewed (changes requested):
Looks mostly good, but missing documentation (and I have a few more smaller comments).

I also want to get @vladzcloudius's review as the expert in tracing.
Vlad, note that the decision here was to **not** add a "table" column to the tracing table (like we did for the "slow query" tracing table). Instead, this patch adds a "table" element in the existing "parameters" column which is a map and today contains just (?) "query".

nyh commented on test/alternator/test_tracing.py:
```diff
@@ -232,19 +234,19 @@ def test_tracing_all(with_tracing, test_table_s_isolation_always, dynamodb):
+    expect_tracing_events(dynamodb, p_getitem, ['GetItem', 'Querying is done'], table.name)
+    expect_tracing_events(dynamodb, p_deleteitem, ['DeleteItem', 'CAS successful'], table.name)
+    expect_tracing_events(dynamodb, p_updateitem, ['UpdateItem', 'CAS successful'], table.name)
+    expect_tracing_events(dynamodb, p_batchgetitem, ['BatchGetItem', 'Querying is done'], table.name)
```
I think we set up tracing for every DynamoDB API request, in alternator/server.cc, server::handle_api_request(). Those other requests also have an alternator_op and a query (the JSON of the request), but wouldn't have a table. 

Please note that we're talking here about **tracing**, not **audit**.

I don't know what we do on requests like CreateTable or DeleteTable - what their tracing looks like (do we see some interesting stuff regarding raft and so on in that trace), and whether we added a "table" parameter for these commands or forgot to do so, and if anybody even cares.

radoslawcybulski replied:
In tracing we add `table` or `table[...]` only if `tracing::add_alternator_table_name` or `tracing::add_table_name` is called, so we should be fine. Test checks only table-related commands, so this should be safe as well.

radoslawcybulski commented on tracing/trace_state.cc:
```diff
@@ -114,6 +114,7 @@ void trace_state::build_parameters_map() {
 
     auto& params_map = _records->session_rec.parameters;
     params_values& vals = *_params_ptr;
+    auto &tables = _records->session_rec.tables;
```
Updated.

nyh commented on tracing/trace_state.hh:
```diff
@@ -501,6 +501,7 @@ private:
     friend void add_prepared_statement(const trace_state_ptr& p, prepared_checked_weak_ptr& prepared);
     friend void set_username(const trace_state_ptr& p, const std::optional<auth::authenticated_user>& user);
     friend void add_table_name(const trace_state_ptr& p, const sstring& ks_name, const sstring& cf_name);
+    friend void add_alternator_table_name(const trace_state_ptr& p, const sstring& cf_name);
```
Please use `std::string_view`, not `const sstring&`. This is always good advice, but even more so in Alternator which often uses std::string, **not** sstring, so by using `sstring &` you are forcing the compiler to make a copy.

nyh commented on tracing/trace_state.hh:
```diff
@@ -677,6 +678,12 @@ inline void add_table_name(const trace_state_ptr& p, const sstring& ks_name, con
     }
 }
 
+inline void add_alternator_table_name(const trace_state_ptr& p, const sstring& cf_name) {
```
The implementation makes me wonder if we shouldn't just call it add_table_name (without the word "alternator") - exactly like the name of the method you call.
The parameter could be called full_table_name instead of "cf_name" - again like in the method that you call.
And as I mentioned above it should take std::string_view, not a sstring reference.

nyh commented on test/alternator/test_tracing.py:
```diff
@@ -229,32 +239,57 @@ def test_tracing_all(with_tracing, test_table_s_isolation_always, dynamodb):
-    assert found_truncated and found_not_truncated
+        elif entry['request'] == op_2:
+            if p_batchwriteitem1 in p and p_batchwriteitem2 in p:
+                tables = entry['parameters']
```
Above you already set p = entry['parameters'], so why do you want to set "tables" to the same thing?

nyh commented on test/alternator/test_tracing.py:
```diff
@@ -229,32 +245,72 @@ def test_tracing_all(with_tracing, test_table_s_isolation_always, dynamodb):
-            p = entry['parameters']
+    batch_get_item_double_tables_session_id = None
+    batch_write_item_double_tables_session_id = None
+    for entry in iterate_tracing_sessions_with_table_filter(dynamodb, table.name):
```
I think I asked this before but can't find my comment so I'll have to write a new one:

What is the relevance of the "table.name" in this call? The code below actually checks for two table names, not one.
As I suggested several times above, I think you didn't need to modify iterate_tracing_sessions() to look for a specific table. I think you could have left the loop over iterate_tracing_sessions() before this patch.

radoslawcybulski replied:
It's a restriction. I'm testing things, so i want to test as much as i can. Restricting sessions to table name makes sure the table name is set and is set correctly. In normal case i'd not bother, but here this is a new functionality (i just added table name to tracing), so i want to check it on every possibility.

radoslawcybulski replied:
You get old behaviour.

nyh reviewed (changes requested):
Thanks, looks better, but there are still multiple questions/comments I had in the previous version that you didn't address.

nyh commented on test/alternator/test_tracing.py:
```diff
@@ -229,32 +239,57 @@ def test_tracing_all(with_tracing, test_table_s_isolation_always, dynamodb):
-                    break
-    assert found_truncated and found_not_truncated
+        elif entry['request'] == op_2:
+            if p_batchwriteitem1 in p and p_batchwriteitem2 in p:
```
I just realized that this is related to my previous question - I think we have p as one long string instead of a dict, so "xyz in p" can find xyz in any one of the different parts of the params - including the "query". So I guess that's what happening here...
But why isn't p a dictionary?

nyh replied:
I'm resolving this conversation. We can't fix it now because of #26870

radoslawcybulski replied:
Yes, you're right, i got there once. Namely to this:

```
static sstring map_to_string(const std::vector<std::pair<data_value, data_value>>& v, bool include_frozen_type) {
    std::ostringstream out;

    if (include_frozen_type) {
        out << "(";
    }

    fmt::print(out, "{}", fmt::join(v | std::views::transform([] (const std::pair<data_value, data_value>& p) {
        std::ostringstream out;
        const auto& k = p.first;
        const auto& v = p.second;
        out << "{" << k.type()->to_string_impl(k) << " : ";
        out << v.type()->to_string_impl(v) << "}";
        return std::move(out).str();
    }), ", "));

    if (include_frozen_type) {
        out << ")";
    }

    return std::move(out).str();
}
```

it's not a json unfortunately.

radoslawcybulski commented:
> Thanks, looks better, but there are still multiple questions/comments I had in the previous version that you didn't address.

Can you please give me links to not addresses comments? I didn't agree with some of them, that's true, but i addresses them all.

nyh commented on test/alternator/test_tracing.py:
```diff
@@ -171,90 +180,144 @@ def expect_tracing_events(dynamodb, str, expected_events):
-    # we will fetch tracing sessions based on operation and beginning of the long unique string
-    # we have set max trace output size to the length of the string, so we can expect truncation
-    # to happen at the end of the string and the beginning should be present
+    with scylla_config_temporary(dynamodb, 'alternator_max_users_query_size_in_trace_output', str(1024 * 32)):
```
You still didn't explain to me why this configuration is helpful in anyway. It only inserts a test for alternator_max_users_query_size_in_trace_output inside a test that was NOT about it. Also note that this test used to pass before we invented alternator_max_users_query_size_in_trace_output. So why do it?

radoslawcybulski replied:
> You still didn't explain to me why this configuration is helpful in anyway. It only inserts a test for alternator_max_users_query_size_in_trace_output inside a test that was NOT about it. Also note that this test used to pass before we invented alternator_max_users_query_size_in_trace_output. So why do it?

Because previously the max size was fixed and everyone was relying on `oh, it's 1024 bytes, i've space`. Not it's not fixed and soon everyone will be `let's add more tracing output, because it might help and if someone doesn't care, he will reduce maximum, because he can` and in 5 years the test will break and someone else will come and will add exactly the code i'm trying to add.
I will do it in a different patch.

radoslawcybulski replied:
> But to tell the truth, I don't want to configure it to be high! I want to see that the default (low) value works as expected.

Not sure what does it mean `as expected`? There's a test that truncation works - i added it. Or do you want to add a test, that default setting is still sensible as doesn't truncate for small operations?

radoslawcybulski replied:
I've removed the call to `scylla_config_temporary`.

radoslawcybulski commented on test/alternator/test_tracing.py:
```diff
@@ -229,32 +250,69 @@ def test_tracing_all(with_tracing, test_table_s_isolation_always, dynamodb):
+    expect_tracing_events_with_session(dynamodb, batch_get_item_double_tables_session_id, ['BatchGetItem', 'Querying is done'])
+    expect_tracing_events_with_session(dynamodb, batch_get_item_double_tables_session_id, ['BatchGetItem', 'Querying is done'])
+    expect_tracing_events_with_session(dynamodb, batch_write_item_double_tables_session_id, ['BatchWriteItem', 'CAS successful'])
+    expect_tracing_events_with_session(dynamodb, batch_write_item_double_tables_session_id, ['BatchWriteItem', 'CAS successful'])
```
Fixed.
